### PR TITLE
Makefile refactoring to factor out common build code

### PR DESCRIPTION
**Pull request by keeranroth**
# Description
Following on from pull request https://github.com/oneapi-src/oneDAL/pull/2614 to add Arm as a supported CPU platform, we update the Makefile structure to make it easier to add other platforms in the future.

The motivation behind this change is to have a minimal set of variables to change when adding a new platform. We also look to re-use makefile code as much as possible in the existing build system, so that it's less likely for builds on different platforms to start to diverge.

The changes to the makefiles require no changes to the build commands.

Most of the changes are copy and paste, with some changes being made to allow the evaluation of make commands to be deferred until all variables required are in scope. No new functionality is added, and no functionality is removed.

Changes proposed in this pull request:
- Add new makefiles to keep platform, arch and uarch specific make variables
- Add new makefiles to keep compiler and backend specific make variables

### CODE REVIEW

File: dev/make/compiler_definitions/clang.ref.32e.mk
Comment (napetrov) [inline dev/make/compiler_definitions/clang.ref.32e.mk:3]:
```suggestion
# Copyright contributors to the oneDAL project
```

File: dev/make/compiler_definitions/gnu.32e.mk
Comment (napetrov) [inline dev/make/compiler_definitions/gnu.32e.mk:2]:
```suggestion
# Copyright contributors to the oneDAL project
```

File: dev/make/compiler_definitions/gnu.mk
Comment (napetrov) [inline dev/make/compiler_definitions/gnu.mk:2]: # Copyright contributors to the oneDAL project

Comment (napetrov) [inline dev/make/compiler_definitions/gnu.mk:2]:
```suggestion
# Copyright contributors to the oneDAL project
```

File: dev/make/function_definitions/32e.mk
Comment (napetrov) [inline dev/make/function_definitions/32e.mk:1]: Should we rename this to be aligned with ARM? 

This is quite a legacy name as one of variants between 32bit, em64t and itanium64 variants.  

Any ideas on what this would be?  basically this is generic x86 architecture

File: dev/make/function_definitions/mac32e.mk
Comment (napetrov) [inline dev/make/function_definitions/mac32e.mk:34]:
```suggestion
endef

```

File: dev/make/function_definitions/32e.mk
Comment (rakshithgb-fujitsu) [reply dev/make/function_definitions/32e.mk:1]: how about x86_64.mk?

Comment (keeranroth) [reply dev/make/function_definitions/32e.mk:1]: Agreed that this could be a better name, and x86_64 sounds decent. However, changing this requires updates outside of the make setup (there are ci scripts that depend on this value), and I think it would be better done as another merge request to change the name. What do you think? I'm happy to put something up for that after this.

Comment (napetrov) [reply dev/make/function_definitions/32e.mk:1]: yes, x86_64 looks decent. - the only thing to check is _ handling and this would depend on how wide we would like to have this implemented, but in longerterm worth having better name.

yes, it's fine to keep this separate as renaming would require updates for internal CI. @homksei @ethanglaser  - this is something that we can look on and rename legacy 32e naming to have better match/separation in future.

Comment (keeranroth) [reply dev/make/function_definitions/32e.mk:1]: https://github.com/oneapi-src/oneDAL/pull/2674 as a draft merge request, for what this would look like on top of this commit.

Is there anything else that needs changing on my end for this merge request? I can see that all the public CI passes, but the Intel CI is failing

File: dev/make/compiler_definitions/gnu.32e.mk
Comment (napetrov) [inline dev/make/compiler_definitions/gnu.32e.mk:2]:
```suggestion
# Copyright contributors to the oneDAL project
```

Current copyright check expects no year defined and i think this is what was suggested by UXL. 
same true for other files

**Comment by keeranroth** [comment]
Looks like another error. Are there some files that I missed the license text on, or is there something else?

**Comment by ethanglaser** [comment]
> Looks like another error. Are there some files that I missed the license text on, or is there something else?

All (6) new files in function_definitions/ directory are failing copyright check. Will look into why this may be happening. Other copyright checks are passing (including files in compiler_definitions/)

**Comment by keeranroth** [comment]
Hi @ethanglaser thanks for having a look at those. Is there anything that I can do on my side? 

Also, is it worth putting an updated copyright check into upstream as well? It would make it easier for me next time round to check this locally (I usually defer to CI checks before pushing a merge request)

**Comment by ethanglaser** [comment]
Hi @keeranroth, it appear that the `Copyright 2024 UXL Foundation` is triggering the internal CI fail

**Comment by ethanglaser** [comment]
Good idea about adding copyright check to public CI - we'll discuss as a team and keep you updated.

Apologies on if I missed discussion on this but did we agree on adding `Copyright 2024 UXL Foundation` to top of new files? If so we should be able to accommodate this internally, otherwise can remove it and checks should pass.

**Comment by keeranroth** [comment]
Ah, my bad on the copyright checks. @napetrov did say that the agreed style doesn't have a year in the statement, so I've updated those

File: makefile
Comment (ethanglaser) [inline makefile:526]: is ARCH needed to be passed? I don't see use of $1 in .compile.template.ay

File: dev/make/compiler_definitions/clang.32e.mk
Comment (ethanglaser) [inline dev/make/compiler_definitions/clang.32e.mk:22]:
```suggestion
#  clang.mkl.32e.mk)
```

File: dev/make/compiler_definitions/clang.mk
Comment (ethanglaser) [inline dev/make/compiler_definitions/clang.mk:22]:
```suggestion
#  clang.32e.mk)
```

**Review by ethanglaser — APPROVED** [review]
Really nice work here, overall just minor questions/comments. Thanks for improving makefile organization! And CI is in good shape

**Review by keeranroth — COMMENTED** [review]
Thanks for the reviews @ethanglaser , and sorry for not having seen this before. Must have missed the emails with your comments in.

I've left one point unaddressed regarding single vs double brackets in a bash script. Whilst I also have a preference for the double brackets, I changed this to match the existing style in the script. Anything else, let me know

File: deploy/local/vars_lnx.sh
Comment (ethanglaser) [reply deploy/local/vars_lnx.sh:1]: looks good

File: dev/make/function_definitions/lnx32e.mk
Comment (ethanglaser) [reply dev/make/function_definitions/lnx32e.mk:19]: yeah good point, this works for me

**Pull request merged**
